Support for Arbitrarily Big Integers

Hirevo opened the pull request:
This PR aims to add support for arbitrarily big integers to `som-interpreter`.  
The support for big integers would be backed by the `BigInt` type from the `num-bigint` crate.  
These take the form of a new variant in the `Value` enum (called `BigInteger`).  
The SOM class associated to such values stays `Integer`, and whether an integer value is stored as `Value::Integer` or `Value::BigInteger` should be abstracted away from SOM programs.  

Any operation on `Value::Integer` that would make it overflow should trigger an automatic promotion to `Value::BigInteger`.  

Some questions remaining unclear:
- Should `Value::BigInteger` ever be downgraded to `Value::Integer` when possible ?
  - And where should that happen ?
- What's the interaction with the already implemented `Value::Integer` to `Value::Double` automatic promotion mechanisms ?
  - Should we implement a `Value::BigDecimal` to allow promoting `Value::BigInteger` to an instance of `Double` ?

smarr commented:
Nice.

> Should Value::BigInteger should ever be downgraded to Value::Integer when possible?
See the use of `asSNumber`: https://github.com/SOM-st/som-java/blob/master/src/som/vmobjects/SBigInteger.java#L56

> What's the interaction with the already implemented Value::Integer to Value::Double automatic promotion mechanisms?

Hm, what do you mean with automatic promotion?
Don't think there's promotion as in 'a widening of value range'.
However, there's a conversion when combining integers and doubles, which should not be affected by BigInteger support.
Note that most SOM implementations are a little sloppy, and don't implement all possible cases.

So, there might be gabs for interactions of big integers and doubles.

Hirevo commented:
> However, there's a conversion when combining integers and doubles, which should not be affected by BigInteger support.

Yes, sorry, it's true that conversion is the better word here, since the max value for a normal `Value::Integer` (which uses a 64-bit signed integer) gets approximated when treated as a `Value::Double` (which uses a 64-bit floating point value), so "promotion" is not accurate:

```plain
(0) SOM Shell | 9223372036854775807
returned: 9223372036854775807 (Integer(9223372036854775807))
(1) SOM Shell | 9223372036854775808
ERROR: could not fully parse the given expression
(2) SOM Shell | 9223372036854775807.0
returned: 9223372036854776000 (Double(9223372036854776000.0))
```

The main question about the conversions is what happens when, for example, a `Value::BigInteger` is added to a `Value::Double` ?  
The big integer cannot really be made into a double since it could be too big.  

> Note that most SOM implementations are a little sloppy, and don't implement all possible cases.
> So, there might be gabs for interactions of big integers and doubles.

Yeah, I think I'll be doing conversions as "best effort" to make most things work and emit an interpreter exception for the cases that cannot be handled properly.

smarr commented:
> The main question about the conversions is what happens when, for example, a Value::BigInteger is added to a Value::Double ?

I think most SOMs do not implement all these cases. So, it's fine to skip over them.

> Yeah, I think I'll be doing conversions as "best effort" to make most things work and emit an interpreter exception for the cases that cannot be handled properly.

I hope the tests tell you what is truly needed, and then there are the benchmarks, which don't use big integers, if I recall correctly.

ltratt commented:
yksom's integer tests -- though almost certainly incomplete -- might be valuable for flushing out unsupported / incorrect cases for this PR, because there are a surprising number of cases. You should feel free to incorporate any of the `lang_test`s from yksom into this repo -- they're all under dual Apache-2.0/MIT, so you can adapt and include them legally! And if you find bugs / unsupported cases, I will happily accept PRs against yksom :)

Hirevo commented:
Sorry for the sudden silence from my part.
Thank you for your proposition of incorporating the tests. I think I'll try to get a script to run the yksom's lang_tests against som-rs every now and then.  
But I don't think I necessarily need to incorporate it, keeping these tests in sync with the rapid pace at which yksom adds new ones (which is a good thing) would be tricky.

Hirevo commented:
About the CI currently failing on this PR:  

This is caused by the `Object>>#hashcode` primitive being somewhat broken.  
I know the reason why but I need to figure out how to properly fix it.  
An issue has been created to fix this: **#4**.

ltratt commented:
> Thank you for your proposition of incorporating the tests. I think I'll try to get a script to run the yksom's lang_tests against som-rs every now and then.
But I don't think I necessarily need to incorporate it, keeping these tests in sync with the rapid pace at which yksom adds new ones (which is a good thing) would be tricky.

Obviously it's completely up to you, but because SOM doesn't specify a lot of things, yksom's tests don't always follow the SOM standard (e.g. SOM doesn't have exceptions/backtraces, but yksom does, so yksom has to "make up" acceptable error messages). That means you'll sometimes want to alter the tests. Personally I run the tests all the time (generally as soon as my current change compiles), and I think that's where their greatest value is, because they catch small problems.

There's also a separate factor which favours copying lang_tests into this repo: not all of them will pass yet because you haven't implemented enough stuff. It can be useful to only have those which you've implemented included.

Anyway, the option is yours, I just wanted to outline some possibilities and emphasise that you're allowed to do whatever you want with yksom's code :)

Hirevo commented:
This PR should be ready to merge now.  
Most operations are now updated to take advantage of big integers and we've been able to fix a few previous inefficacies with this feature.  
(the only remaining ones have to do with right-shifting, but I do not think it should block merging the PR)
So, I've rebased this PR on latest master in preparation to merge, after the CI run.

The benchmark CI workflow fails because this branch misses the changes from **#8**, so it should be all good after both PRs get merged.
